### DISCUSSION TITLE
Pull Request Diffs Do Not Update After Changes Are Made

### DISCUSSION THREAD

**Discussion by adammfrank** [community/community#16351] [Pull Requests]
When I comment on a PR requesting a change and the change is made, the diff does not show the change. It only shows the file as it was when I made the comment. Even if I click "view file" it shows the old one. That means, in order to verify that the change satisfied my request, I need to leave Github, pull down the branch and find it myself. All while Github tells me that file had "Changed since last view" without showing me the changes.

See previous discussion here: https://github.com/isaacs/github/issues/750#issuecomment-872563253

That Issue suggested posting here in 2021, but it doesn't look like anyone did.

**Comment by jeacott1** [comment]
pull request diffs seem to be using a 3 dot diff instead of a 2 dot diff ( 2 dot diff against the current head of the target branch is what it should be doing.)
if I create a branch and a PR from it to main, then update main, my PR remains showing a compare against the original branch base. 
What's worse is that sometimes github even fails to show a merge conflict where one absolutely exists. it used to be that switching target branches in the github PR and then back again would wake up the diff, but alas this no longer works. the ONLY thing that works is rebasing your branch on the updated main and force pushing the branch to github again.

its dangerous and it makes the github pull request/ review screen completely useless. they need to fix it.

**Reply by numberZero** [reply]
> 2 dot diff against the current head of the target branch is what it should be doing

No, no and no. It should do a 2-dot diff between target branch and *merge result*. Which is essentially what a 3-dot diff emulates, sometimes well, sometimes poorly.

**Reply by jeacott1** [reply]
@numberZero a 3 dot diff wont show you the difference against the head of the target branch, only the changes since the branch was made, and that's the problem here.
I agree that what I think we actually want here is target branch vs merge result, but a 3 dot diff doesn't get you anywhere close to that.

**Comment by th-anguyen** [comment]
Yeah I'm seeing the same thing here.  Open the PR, make changes, and the PR `Files Changed` diff tab doesn't reflect the latest changes.  Stupid and confusing as reviewers can't really re-review from the Github PR page.

**Reply by crbertoldo** [reply]
Do you happen to know whether this has been classified as a bug and is tracked somehow? I couldn't find it...

**Comment by ihtisham-sohail** [comment]
I am also facing this issue , even I removed the changes and added them again with new commits , still it's not detecting them.
The commit is showing , even the changes are shown when I click on that commit , but it isn't showing up in the default changes view.
Also , just for checking , I pull the changes into the base branch ( The changes that are not being detected ) and yes they are being merged, just not showing up in the review changes tab.

**Comment by uherm-anz** [comment]
Yeah, this is making reviews more difficult than they need to be. Would love for this to be addressed

**Comment by ClementWalter** [comment]
the doc is however explicit about this decision https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#three-dot-and-two-dot-git-diff-comparisons

**Reply by JacekCala** [reply]
Yeah, I see the need to understand what the PR introduced, but would be great to have a checkbox to be able to easily compare with the latest commit from the base branch. I wonder if this is possible in any near future.

**Reply by jeacott1** [reply]
for me this decision makes no sense at all. a PR is clearly a compare and merge against a target 'branch', PR's are not about the changes on a single branch - you can see that info anywhere.

**Reply by bennyvdhoogen** [reply]
Ran into this today, having worked with Gitlab for the last couple of years this really confuses me. From my experience the Gitlab UI updates Merge Request automatically whenever the base branch is updated. In this process commits that have been merged in the mean time are left out of the view and only the code changes 'at present' are displayed. 

Not sure if both these differences are represented just in the diffing methods used, but I suppose Github favours 3 dot diffs whereas GitLab favors 2 dot diffs.

**Reply by numberZero** [reply]
2-dot comparison between base and topic branches makes no sense at all. It would show lines removed from the base branch as additions which is clearly wrong, the actual merge won’t readd them. 3-dot comparison is better as it shows the changes to be merged. But there is a third option: 2-dot comparison between base branch and *merge result.* I.e. the changes that would *actually* be introduced by the PR. But it is way more expensive to compute as it requires actually performing the merge just to show the diff, and may also fail if there are conflicts.

Also, what the OP describes sounds more like a bug than 2-or-3-dot choice. Something along the lines of https://github.com/orgs/community/discussions/51948.

**Reply by 0xCooki** [reply]
> Yeah, I see the need to understand what the PR introduced, but would be great to have a checkbox to be able to easily compare with the latest commit from the base branch. I wonder if this is possible in any near future.

Yes, this please

**Comment by arunkant** [comment]
This is a shock to our development team migrating from bitbucket recently. Any workaround I can implement to show diff from target branch and not all the changes done since last common commit? I want to see the changes a PR will do

**Reply by ihtisham-sohail** [reply]
Well I did find a way to compare the branches in 2 dot comparison ( by default it's in 3 dot when a PR is created ), not sure it might help your case but it might give you another depth in the changes between the 2 branches.
In your Git URL when comparing the 2 branches , you can change the comparison method by changing the number of preceding dots. For example :-

`https://github.com/your-project-name/project-folder>/compare/branch-1...branch-2`

change the 3 dots to 2 dots and you'll start viewing it in 2 dot comparison

`https://github.com/your-project-name/project-folder>/compare/branch-1..branch-2`

![Git 2 dot branch comparison](https://user-images.githubusercontent.com/61691880/234530119-83fe7d4e-6d67-4528-b8c0-04842aa217b3.jpg)

**Reply by electricenjindevops** [reply]
You sir are a hero to many.

**Comment by yuanzhou** [comment]
I've noticed this a while ago and couldn't figure out the cause. Sometimes a PR also shows conflicts but when I did the manual merge there's none. Also sometimes when we update the base branch of PR in hope of resolving the conflicts, the PR page still shows the same conflicts even though in fact the conflicts have been resolved due to the latest commit made into the base branch.

**Reply by yuanzhou** [reply]
I tested a few times, sometimes even with manually changing to the two dot diff in the URL while doing a code comparison between two branches, I don't get the actual changes showing correctly.

**Comment by ms-ati** [comment]
This is happening to me today -- PRs are not updating after pushing to the branch

**Reply by hugosjoberg** [reply]
Same here

**Reply by thebitspud** [reply]
Also having this issue, updates were working fine earlier today 😐

**Reply by geovanygameros** [reply]
They are fixing it https://www.githubstatus.com

**Reply by RajaAsyraf** [reply]
having same issue, hopefully they can fix it soon 🤞🏼

**Reply by lamontadams** [reply]
wow. why is this still an issue?

**Reply by pyaf** [reply]
I'm also facing the issue. This is very frustrating, not sure how to fix this.

**Reply by CarrKaden** [reply]
Facing this issue as well

**Reply by RafSobol** [reply]
Same issue happens to myself and another user in the organisation (on 2 separate PRs, but that's 2 out of 2). All green according to githubstatus.com 😢

**Reply by ayushxx7** [reply]
facing the same. as a workaround, changing the `base` branch to some other branch, and switching back to original branch fixes the issue.

**Reply by aadypillai-hive** [reply]
this worked for me ^

**Reply by nighto** [reply]
Having the same issue today, while githubstatus.com shows all green

**Reply by sahilnarwal-nxt** [reply]
+1

**Reply by rhpvorderman** [reply]
Having the same issue on https://github.com/rhpvorderman/sequali/pull/18

**Reply by bongolan** [reply]
+1

**Reply by shivaraj-bh** [reply]
+1
Having the issue here https://github.com/juspay/services-flake/pull/7

**Reply by simas2275** [reply]
+1

**Reply by zharfan104** [reply]
+1

**Reply by fadi-quader-mox** [reply]
➕ 1️⃣

**Reply by grisharav** [reply]
+1

**Reply by mukeshpanchal27** [reply]
+1

**Reply by AlexPan1992** [reply]
+1

**Reply by varadkarpe** [reply]
+1

**Reply by nighto** [reply]
they seem to have acknowledged the issue: https://www.githubstatus.com/incidents/pr0ptqcw6q8d

**Reply by zbynekwinkler** [reply]
Changing base branch to something different and then back to the original fixed the problem for me - now all commits can be seen in the PR.

**Reply by nighto** [reply]
on the issue description they say that simply making a new commit to the branch is enough to fix the PR.

**Reply by MasoudBimar** [reply]
+1

**Comment by ayushxx7** [comment]
seems like this is a nice workaround: https://github.com/orgs/community/discussions/16351#discussioncomment-6327770
> changing the base branch to some other branch, and switching back to original branch fixes the issue.

[wanted to make the solution easily visible 😄]

**Reply by jeacott1** [reply]
this hasn't worked for me for a very long time. It did work years ago.

**Reply by dmosorast** [reply]
Just ran into this for the first time, and this fixed the view in the PR for me. Thanks!

**Reply by znemeth23** [reply]
yeah, sadly it does not work in case of a newly created PR :(

**Comment by grisharav** [comment]
Also I think the actions are not being run on the missing commits even after changing the base branches... It's a major issue.

**Comment by alex-djangsters** [comment]
It's still an issue, diff is strange, merge conflicts, etc.
Only local rebase and force push works for me.

**Comment by jeremymartin-snowden** [comment]
Coming from Azure Devops this makes no sense to me.. I get that some people might prefer it so make it an option against the repository?

**Reply by jeacott1** [reply]
this is how devops works too by default, although it looks like there's an option to at least force it to update.
https://learn.microsoft.com/en-us/azure/devops/repos/git/complete-pull-requests?view=azure-devops&tabs=browser#check-merge-changes

**Comment by shivvamm** [comment]
It seems you're experiencing a problem with GitHub's pull request (PR) interface not displaying the correct changes after a requested change has been made. This can be frustrating, especially when trying to verify changes without leaving GitHub. To address this:

Clear Browser Cache: Sometimes, caching issues can cause discrepancies in the displayed content. Try clearing your browser cache and refreshing the PR page.

Hard Refresh: Use a hard refresh (usually Ctrl + F5 or Shift + Ctrl + R) to force your browser to fetch the latest content.

Review Comments: Make sure the PR author has properly addressed your comments. Review the PR discussion to confirm that the requested changes were indeed made.

Use GitHub's "Review Changes" Feature: Instead of relying solely on viewing individual files, use GitHub's "Review Changes" feature. It displays all changes in the PR at once, making it easier to verify changes.

Compare View: In the PR, click on the "Files changed" tab and then click "Review changes." This view should show you all the changes made in response to your comments.

Diff URL: Manually check the diff URL in the address bar. It should include the commit hash of the latest changes. For example: https://github.com/user/repo/pull/123/files?diff=split#commit-hash.

Logged Out: Sometimes, staying logged out of GitHub and then logging back in can help resolve display issues.

Check GitHub Status: Occasionally, GitHub may experience service disruptions. Check the [GitHub Status](https://www.githubstatus.com/) page to ensure there are no ongoing issues.

Contact GitHub Support: If the problem persists, you can reach out to [GitHub Support](https://support.github.com/) for assistance. Use the GitHub Support platform to report the issue.

Stay Updated: Keep an eye on the discussion in the GitHub issue you mentioned. Others might share solutions or workarounds if they've experienced the same problem.

Remember, while waiting for a resolution, you can continue using the workaround of pulling down the branch to verify changes if needed.

**Reply by jeacott1** [reply]
this isn't a refresh/browser cache issue.

**Reply by znemeth23** [reply]
hmmm, why does this answer sound exactly like an AI-generated one? 🤔

**Comment by shivvamm** [comment]
sorry for my misunderstanding

**Comment by olblak** [comment]
Sharing my experience here, as it's hard to find information on this issue, and it's really bother me

I am maintaining a declarative dependency management tool to automate file update https://github.com/updatecli/updatecli/
When a file needs to be updated, it creates a temporary branch and open a pull request

For various reason, Updatecli will often reuse the same temporary branch so we automatically delete them when a pull request is merged.
This used to work great until end of Mars 2023 where it became totally broken for a few weeks, this resulted in thousands of empty pull request on our projects.
For those using auto merge, it resulted in thousands of empty commits.

* https://github.com/updatecli/updatecli/issues/1252

It took us a few weeks back then to recover from it, and since then we randomly see wrong "diff" information like from an old branch which is supposed to be have been delete since then.

An example here https://github.com/civo/kubernetes-marketplace/pull/545#discussion_r1303929371

Usually we just close the pull request, delete the branch, and let Updatecli recreate everything correctly but ideally I would like to find a solution

**Comment by jraczykowski** [comment]
Why github is so stubron  on basic functionalities? The same goes with fast-forward merges. So baSIC!

**Comment by aamir747** [comment]
Faced this today and the only thing that worked for me is to create a fresh PR, merge my previous PR into it to get the correct diff against main

**Comment by xharriscisco** [comment]
I've been having this issue for months. Re-selecting the base branch (ie. main) for a PR usually forces it to update the file changes diff.

Edit: I contacted github support and they said it's a known quirk and they have an open issue to fix it in a future release.

**Reply by ryanisn** [reply]
thanks for the trick, this has been bother me for years

**Comment by jasells** [comment]
> Why github is so stubron on basic functionalities? The same goes with fast-forward merges. So baSIC!

Because github is trash, apparently.  This is, indeed basic functionality, and if it can't get this right why should I trust it to do anything?

**Reply by jeacott1** [reply]
It's a Microsoft thing. Azure Devops has exactly the same problem, by design.

**Reply by jasells** [reply]
Being an ADO user daily for >6 years, I can say I have NEVER seen anything like a PR "lag" on showing updates.  
To update my personal status, the PR did finally update eventually.  Not sure what caused it or how long it took exactly, since I got pulled away on other things, and then when I came back, to it is was showing the new commit's changes, and showing a new commit on the PR's summary page.

**Reply by jeacott1** [reply]
This discussion isn't about any kind of lag, it's about pull requests comparing the PR feature branch head with the commit on the target branch from where the PR branch was created - this is NOT how PR's should work. A PR SHOULD compare the head of both the feature and target branches, but by design both github and devops don't work like this, and the only fix is to rebase your feature branch and push again.

**Reply by jasells** [reply]
_Technically_ the OP is talking about a comment on a PR "perma-linked" to the original code, not the updated code in response to the comment, in the feature branch.  Which is still an issue, different from not showing commits on target OR feature at all.  ADO works great in this regard, in that the default behavior is to show the latest code @ the line the comment is attached, but there's a btn that shows and allows you to "see comment in the original context." Super easy to review and make sure changes from review feedback are complete.

My issue with GH that led me here was that new commits on a _feature_ branch were not showing in the pr, though I could see them on the branch itself.  Possibly, I noticed the leading edge of some service interruptions GH had this week, since the commits eventually did show up in the PR and the diffs (15-30 minutes or more)?

As far as rebasing the feature branch or otherwise syncing with the original branch, I haven't noticed the issue often on ADO, likely because my SOP is to rebase/sync a feature branch before opening a PR, mainly to resolve conflicts.  From there, yes, ADO will at least show when the target has updated, and _usually_ auto-re-merge, I think, but at least you get a button right @ the top of the PR to re-start the merge in the latest version.  It may be dependent on policies, like if there's a policy-build/pipeline may auto-remerge on feature-branch update.

**Comment by 5pecia1** [comment]
I closed and reopened the PR, and then the diff on GitHub worked properly.

**Comment by simchavos** [comment]
Wondering if this will be addressed at some point since it can easily lead to unwanted behaviour / wrong code reviews..

**Comment by AngusMaiden** [comment]
Also just discovered this issue after 5 years of using GitHub professionally.  I was wondering why I was not seeing the right diff after the target branch updated and a Google search led me here.

I didn't even notice before, as it's a silent error. I thought that what I was always seeing was "Files changed" between the target branch and incoming branch, so I never questioned it until it became very obvious in my latest PR. It is absolutely not expected behaviour for it to be anything else. And yet it is.

Are GitHub ever going to address this really poor user experience?

**Reply by znemeth23** [reply]
agreed, it's ridiculous that such a basic functionality is broken for years...
GitHub's "backward logic", calling merge requests "pull requests", etc. is frustrating on its own (especially coming from a Gitlab background), but this PR diff issue is just the cherry on top...

**Comment by DenyVeyten** [comment]
Bitbucket had a correct diff showing before their "performance improvement" in 2022 that intentionally dropped the "3-way" diff we all here want https://www.atlassian.com/blog/bitbucket/improving-performance-on-complex-diffs

There was a comment somewhere from the maintainer of that feature that it became too hard to maintain this feature effort-wise, as it was conflicting with other Bitbucket features (which could be a similar cause for GitHub), but I can't find their comment anymore (maybe it was removed/hidden/redacted).

That has become a huge pain for anyone who does not want to follow a rigid gitflow or trunk-based development (i.e. can't simply rebase/pull target branch into the source branch without creating more temporary branches). And I suppose it was always the case for GitHub in particular?
